### PR TITLE
Add slices mode -> fixed number of slices

### PR DESCRIPTION
**Pull request by webrouse**
Jire: https://keboola.atlassian.net/browse/COM-847

Changes:
- Added `slices` mode.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,9 +13,10 @@
 
 It supports optional parameters:
 
-- `mode` - enum `bytes` or `rows`, default `bytes`
+- `mode` - enum (`bytes`, `rows`, `slices`), default `bytes`
 - `bytesPerSlice` (`int`) - for `mode = bytes`, maximum size of the one slice in bytes before compression, default `524 288 000` (`500 MiB`)
 - `rowsPerSlice` (`int`) - for `mode = rows`, maximum rows in the one slice, default `1 000 000`
+- `numberOfSlices` (`int`) - for `mode = slices`, fixed number of slices, default `60`
```
Comment (webrouse) [inline README.md:19]: V povodnom zadani bolo, ze idealne ked CSVcko rozdelime na `60 casti` ... doteraz to nedavalo prakticky zmysel, ... no z testov nad Synapse vyzera, ze Synapse moze pomoct viacero malych casti ... tak to tu pridavam, aby som to zmeral.

File: src/csv/slicedWriter/slicedWriter.go [hunk 2]
```diff
@@ -3,29 +3,65 @@ package slicedWriter
...
-	w := &SlicedWriter{conf: conf, dirPath: dirPath}
-	w.createNextSlice() // open first slice
+func NewSlicedWriterFromConf(conf *config.Config, inFileSize uint64, outPath string) *SlicedWriter {
+	mode := conf.Parameters.Mode
+	bytesPerSlice := conf.Parameters.BytesPerSlice
+	rowsPerSlice := conf.Parameters.RowsPerSlice
+	maxSlices := conf.Parameters.NumberOfSlices
+
+	// If fixes number of slices -> calculate bytesPerSlice
+	if mode == config.ModeSlices {
```
Comment (webrouse) [inline src/csv/slicedWriter/slicedWriter.go:33]: Tu sa prevadza `numberOfSlices` na `bytesPerSlices` ... kedze kazda cast ma byt rovnako velka.

File: src/csv/slicedWriter/slice.go [hunk 3]
```diff
@@ -24,14 +24,14 @@ type slice struct {
 	bytes    uint64
 }
 
-func NewSlice(c *config.Config, filePath string) *slice {
+func NewSlice(mode config.Mode, maxBytes uint64, maxRows uint64, gzipEnabled bool, gzipLevel int, filePath string) *slice {
```
Comment (webrouse) [inline src/csv/slicedWriter/slice.go:27]: Neposila sa tu config ale jednotlive hodnoty, ... kedze v `CreateSlicedWriterFromConf` sa prepocitava `numberOfSlices` na `bytesPerSlice` a tu ide uz vysledok.

File: src/csv/slicedWriter/slicedWriter.go [hunk 4]
```diff
@@ -38,8 +74,17 @@ func (w *SlicedWriter) Close() {
 	w.slice.Close()
 }
 
+func (w *SlicedWriter) IsSpaceForNextRowInSlice(rowLength uint64) bool {
+	// Last slice, do not overflow
```
Comment (webrouse) [inline src/csv/slicedWriter/slicedWriter.go:78]: Nikdy sa netrafime presne, kedze kazdy riadok v CSV ma inu a neznamu dlzku.
A do kazdej `slice` sa uklada iba cely pocet riadkov -> zvysnych par riadkov proste zapiseme do poslednej `slice` ... zabranime tak tomu, aby vznikla posledna `slice`, ktora by mala iba par riadkov/bajtov + vzdy nam bude sediet zadany pocet.

File: src/csv/slicedWriter/slicedWriter_test.go [hunk 5]
```diff
@@ -29,7 +29,7 @@ func TestNewSlicedWriter(t *testing.T) {
 	}
 
 	// Create writer
-	w := NewSlicedWriter(conf, tempDir)
+	w := NewSlicedWriterFromConf(conf, 1000, tempDir)
```
Comment (webrouse) [inline src/csv/slicedWriter/slicedWriter_test.go:32]: `1000` velkost vstupneho suboru, pouziva sa iba pre `mode = slices`.

File: src/csv/slicedWriter/slicedWriter.go [hunk 6]
```diff
@@ -3,29 +3,65 @@ package slicedWriter
...
+	mode := conf.Parameters.Mode
+	bytesPerSlice := conf.Parameters.BytesPerSlice
+	rowsPerSlice := conf.Parameters.RowsPerSlice
+	maxSlices := conf.Parameters.NumberOfSlices
+
+	// If fixes number of slices -> calculate bytesPerSlice
+	if mode == config.ModeSlices {
+		mode = config.ModeBytes
+		fileSize := float64(inFileSize)
+		bytesPerSlice = uint64(math.Ceil(fileSize / float64(maxSlices)))
```
Comment (webrouse) [inline src/csv/slicedWriter/slicedWriter.go:36]: Tu je `fileSize` -> velkost vstupneho suboru, ... pouziva sa iba pre `mode = slices`.

File: src/tests/mode-slices/expected-stdout [hunk 7]
```diff
@@ -0,0 +1,3 @@
+Configured number of slices = 4.
+Slicing table "tables/tenRows.csv".
+Table "tables/tenRows.csv" sliced. Written 4 parts, 10 rows, total size 101 B. Manifest created.
```
Comment (webrouse) [inline src/tests/mode-slices/expected-stdout:3]: Datadir test.

File: src/tests/mode-slices/expected/data/out/tables/tenRows.csv/part0004 [hunk 8]
```diff
@@ -0,0 +1,4 @@
+"7","abc"
+"8","abc"
+"9","abc"
```
Comment (webrouse) [inline src/tests/mode-slices/expected/data/out/tables/tenRows.csv/part0004:3]: Zvysne riadky sa zapisali do poslednej `slice`. ... tu je v ostatnych 2 a v poslednej 4 riadky. No v relanej prevadzke bude rozdiel mensi. Tj. napr. `1000` a `1002`.

File: src/utils/utils.go [hunk 9]
```diff
@@ -56,6 +56,15 @@ func FileExists(path string) bool {
 	return false
 }
 
+func FileSize(path string) uint64 {
```
Comment (webrouse) [inline src/utils/utils.go:59]: Pomocna metoda na ziskanie velkosti suboru.

File: src/tests/mode-slices-empty-table/expected-stdout [hunk 10]
```diff
@@ -0,0 +1,3 @@
+Configured number of slices = 4.
+Slicing table "tables/empty.csv".
```
Comment (webrouse) [inline src/tests/mode-slices-empty-table/expected-stdout:2]: Pridal som este testy pre edge cases: prazdna tabulka, 1 riadok, 2 riadky, .. kod sa nemenil.

**Review by ondrajodas — APPROVED** [review]
